crypto/fipsmodule/service_indicator/service_indicator.c

```diff
@@ -59,40 +60,78 @@ void FIPS_service_indicator_update_state(void) {
   indicator->counter++;
 }
 
-void AES_verify_service_indicator(unsigned key_rounds) {
-  switch (key_rounds) {
-    case 9:
-    case 10:
-    case 11:
-    case 12:
-    case 13:
-    case 14:
-      FIPS_service_indicator_update_state();
-      break;
-    default:
-      break;
+void AES_verify_service_indicator(const EVP_CIPHER_CTX *ctx, const unsigned key_rounds) {
+  if(ctx != NULL) {
+    if (EVP_CIPHER_CTX_mode(ctx) == EVP_CIPH_ECB_MODE ||
+        EVP_CIPHER_CTX_mode(ctx) == EVP_CIPH_CBC_MODE ||
+        EVP_CIPHER_CTX_mode(ctx) == EVP_CIPH_CTR_MODE) {
+      switch (ctx->cipher->key_len) {
+        case 16:
+        case 24:
+        case 32:
+          FIPS_service_indicator_update_state();
+          break;
+        default:
+          break;
+      }
+    }
+  } else {
+    switch (key_rounds) {
+      case 9:
+      case 10:
+      case 11:
+      case 12:
+      case 13:
+      case 14:
+        FIPS_service_indicator_update_state();
+        break;
+      default:
+        break;
+    }
   }
 }
 
-void AEAD_verify_service_indicator(size_t key_length) {
-  switch (key_length) {
-    case 16:
-    case 32:
-      FIPS_service_indicator_update_state();
-      break;
-    default:
-      break;
+void AEAD_GCM_verify_service_indicator(const EVP_AEAD_CTX *ctx) {
+  // Not the best way to write this, but the delocate parser for ARM/clang can't
+  // recognize || if statements, or switch statements for this.
+  // TODO: Update the delocate parser to be able to recognize a more readable
+  // version of this.
+  // We only have support for 128 bit and 256 bit keys for AES-GCM.
+   if(EVP_AEAD_key_length(ctx->aead) == 16) {
+    FIPS_service_indicator_update_state();
+   } else if(EVP_AEAD_key_length(ctx->aead) == 32) {
+    FIPS_service_indicator_update_state();
+   }
+}
+
+void AEAD_CCM_verify_service_indicator(const EVP_AEAD_CTX *ctx) {
+  // Only 128 bit keys with 32 bit tag lengths are approved for AES-CCM.
+  if(EVP_AEAD_key_length(ctx->aead) == 16 && ctx->tag_len == 4) {
+    FIPS_service_indicator_update_state();
   }
 }
 
 #else
 
 uint64_t FIPS_service_indicator_before_call(void) { return 0; }
 uint64_t FIPS_service_indicator_after_call(void) { return 0; }
-int FIPS_service_indicator_check_approved(int before, int after) { return AWSLC_NOT_APPROVED; }
 
-void AES_verify_service_indicator(unsigned key_rounds) { }
-void AEAD_verify_service_indicator(size_t key_length) { }
+
+// Service indicator check functions listed below are optimized to not do extra
+// checks, when not in FIPS mode. Arguments must be cast to void to avoid
+// unused warnings.
+
+int FIPS_service_indicator_check_approved(OPENSSL_UNUSED int before,
```

Comment:
In #263 https://github.com/awslabs/aws-lc/pull/263/files#diff-ce7ffb6368fc741735e07a58ef82c4e0bb9b31d389299f8d0d1e961c76dd9f38R147

this returns `AWSLC_NOT_APPROVED`.